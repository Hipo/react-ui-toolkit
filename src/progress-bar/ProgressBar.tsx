import "./_progress-bar.scss";

import React from "react";
import classNames from "classnames";

export interface ProgressBarProps {
  percentage: number;
  style: {
    trackColor?: string;
    backgroundColor?: string;
    completedColor?: string;
  };
  children?: React.ReactNode;
  customClassName?: string;
  testid?: string;
}

function ProgressBar(props: ProgressBarProps) {
  const {
    percentage,
    style: {trackColor = "blue", backgroundColor = "gray", completedColor = "green"},
    children,
    customClassName,
    testid
  } = props;
  const progressBarClassName = classNames("progress-bar", customClassName);
  // eslint-disable-next-line no-magic-numbers
  const parsedPercentage = percentage >= 100 ? 100 : percentage;

  return (
    <div
      className={progressBarClassName}
      style={{
        backgroundColor
      }}
      role={"progressbar"}
      aria-valuenow={parsedPercentage}
      aria-valuemin={0}
      aria-valuemax={100}
      data-testid={testid}>
      <div
        className={"progress-bar__track"}
        style={{
          width: `${parsedPercentage}%`,
          // eslint-disable-next-line no-magic-numbers
          backgroundColor: parsedPercentage === 100 ? completedColor : trackColor
        }}
        data-testid={`${testid}.track`}
      />

      {children && <div className={"progress-bar__content"}>{children}</div>}
    </div>
  );
}

export default ProgressBar;
