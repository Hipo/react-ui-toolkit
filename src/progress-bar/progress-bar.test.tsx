import React from "react";
import {render, screen} from "@testing-library/react";
import "@testing-library/jest-dom";
import {create} from "react-test-renderer";

import ProgressBar, {ProgressBarProps} from "./ProgressBar";
// import {testA11y} from "../core/utils/test/testUtils";

describe("<ProgressBar />", () => {
  const defaultProgressBarProps: ProgressBarProps = {
    testid: "progress-bar",
    percentage: 57,
    style: {backgroundColor: "blue", completedColor: "yellow", trackColor: "orange"}
  };

  it("should render correctly", () => {
    render(<ProgressBar {...defaultProgressBarProps} />);
  });

  it("should match snapshot", () => {
    const tree = create(<ProgressBar {...defaultProgressBarProps} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  // it("should pass a11y test", async () => {
  //   const {container} = render(<ProgressBar {...defaultProgressBarProps} />);

  //   await testA11y(container);
  // });

  it("should render children correctly", () => {
    const progressBarContent = <span data-testid={"progress-bar.content"}>{"Test"}</span>;

    render(<ProgressBar {...defaultProgressBarProps}>{progressBarContent}</ProgressBar>);

    expect(screen.getByRole("progressbar")).toContainElement(screen.getByText("Test"));
  });

  it("progress bar's background color should be backgroundColor", () => {
    render(<ProgressBar {...defaultProgressBarProps} />);

    expect(screen.getByRole("progressbar")).toHaveStyle(
      `background-color: ${defaultProgressBarProps.style.backgroundColor}`
    );
  });

  it("Progress bar track has the right size", () => {
    const {rerender} = render(<ProgressBar {...defaultProgressBarProps} />);

    expect(screen.getByTestId(`${defaultProgressBarProps.testid!}.track`)).toHaveStyle(
      `width: ${defaultProgressBarProps.percentage}%`
    );

    rerender(<ProgressBar {...defaultProgressBarProps} percentage={120} />);

    expect(screen.getByTestId(`${defaultProgressBarProps.testid!}.track`)).toHaveStyle(
      "width: 100%"
    );
  });

  it("if the percentage is less than 100, background color of the progress bar track should be trackColor", () => {
    render(<ProgressBar {...defaultProgressBarProps} />);

    expect(screen.getByTestId(`${defaultProgressBarProps.testid!}.track`)).toHaveStyle(
      `background-color: ${defaultProgressBarProps.style.trackColor}`
    );
  });

  it("if the percentage is greater or equal to 100, background color of the progress bar track should be completedColor", () => {
    const {rerender} = render(
      <ProgressBar {...defaultProgressBarProps} percentage={100} />
    );

    expect(screen.getByTestId(`${defaultProgressBarProps.testid!}.track`)).toHaveStyle(
      `background-color: ${defaultProgressBarProps.style.completedColor}`
    );

    rerender(<ProgressBar {...defaultProgressBarProps} percentage={120} />);

    expect(screen.getByTestId(`${defaultProgressBarProps.testid!}.track`)).toHaveStyle(
      `background-color: ${defaultProgressBarProps.style.completedColor}`
    );
  });
});
