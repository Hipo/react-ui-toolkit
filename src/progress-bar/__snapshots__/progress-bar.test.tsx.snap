// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressBar /> should match snapshot 1`] = `
<div
  aria-valuemax={100}
  aria-valuemin={0}
  aria-valuenow={57}
  className="progress-bar"
  data-testid="progress-bar"
  role="progressbar"
  style={
    Object {
      "backgroundColor": "blue",
    }
  }
>
  <div
    className="progress-bar__track"
    data-testid="progress-bar.track"
    style={
      Object {
        "backgroundColor": "orange",
        "width": "57%",
      }
    }
  />
</div>
`;
